gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -952,6 +952,10 @@ private void submitJob(DagNode<JobExecutionPlan> dagNode) {
         TimingEvent jobOrchestrationTimer = this.eventSubmitter.isPresent() ? this.eventSubmitter.get().
             getTimingEvent(TimingEvent.LauncherTimings.JOB_ORCHESTRATED) : null;
 
+        if (this.metricContext != null) {
+          getRunningJobsCounter(dagNode).inc();
+          getRunningJobsCounterForUser(dagNode).forEach(ContextAwareCounter::inc);
```

Comment:
this may require a comment (at least for me to grok)... seems you are incrementing counters before actually submitting the job for execution, correct? (even prior to this change we didn't wait to find for the future's result)

the async future does present a challenge (since we don't await fulfillment).  should perhaps the increment happen within the future, once the result of submission is actually know?

I could imagine an argument for using the increment as a sort of pre-reservation (locking) to avoid exceeding the allotment.  therefore doing just prior to submission.  but, that said, I thought quota management (`checkQuota`) does just that, so wouldn't be necessary to duplicate.

anyway... I may be confused... just let me know your strategy here--and consider a comment so I (and other maintainers) will more readily understand.

Reply:
Adding this here for future committers/documentation:

The future isn't actually working in async despite providing the interface, so exceptions are thrown in `addSpec()` function which is misleading. This causes the thread to terminate early so that incrementing is never done, but since the job is marked as FAILED then there will be a corresponding decrement since the job gets allocated a quota already.